api/api.go

```diff
@@ -128,29 +129,79 @@ func New(ac *Config) (*API, error) {
 	return a, nil
 }
 
+// EnableExponentialBackoff enables use of exponential backoff for next API call(s)
+// and use exponential backoff for all API calls until exponential backoff is disabled.
+func (a *API) EnableExponentialBackoff() {
+	a.useExponentialBackoff = true
+}
+
+// DisableExponentialBackoff disables use of exponential backoff. If a request using
+// exponential backoff is currently running, it will stop using exponential backoff
+// on its next iteration (if needed).
+func (a *API) DisableExponentialBackoff() {
+	a.useExponentialBackoff = false
+}
+
 // Get API request
 func (a *API) Get(reqPath string) ([]byte, error) {
-	return a.apiCall("GET", reqPath, nil)
+	return a.apiRequest("GET", reqPath, nil)
 }
 
 // Delete API request
 func (a *API) Delete(reqPath string) ([]byte, error) {
-	return a.apiCall("DELETE", reqPath, nil)
+	return a.apiRequest("DELETE", reqPath, nil)
 }
 
 // Post API request
 func (a *API) Post(reqPath string, data []byte) ([]byte, error) {
-	return a.apiCall("POST", reqPath, data)
+	return a.apiRequest("POST", reqPath, data)
 }
 
 // Put API request
 func (a *API) Put(reqPath string, data []byte) ([]byte, error) {
-	return a.apiCall("PUT", reqPath, data)
+	return a.apiRequest("PUT", reqPath, data)
+}
+
+// apiRequest manages retry strategy for exponential backoffs
+func (a *API) apiRequest(reqMethod string, reqPath string, data []byte) ([]byte, error) {
+	backoffs := []uint{2, 4, 8, 16, 32}
+	attempts := 0
+	success := false
+
+	var result []byte
+	var err error
+
+	for !success {
+		result, err = a.apiCall(reqMethod, reqPath, data)
+		if err == nil {
+			success = true
+		}
+
+		// break and return error if not using exponential backoff
+		if err != nil {
+			if !a.useExponentialBackoff {
+				break
+			}
+		}
+
+		if !success {
+			var wait uint
+			if attempts >= len(backoffs) {
+				wait = backoffs[len(backoffs)-1]
+			} else {
+				wait = backoffs[attempts]
+			}
+			attempts++
+			a.Log.Printf("[WARN] API call failed %s, retrying in %d seconds.\n", err.Error(), wait)
+			time.Sleep(time.Duration(wait) * time.Second)
```

Comment:
If we have lots of machine booting at the same time (or cron'd in their use of the API) this backoffs will align.  Might I suggest:   wait * (1 + rand())/2

This will randomize the back off somewhere between 50% and 100% of the intended backoff schedule.  It will give competing calls (nodes/apps) an opportunity to not collide.